### DISCUSSION TITLE
Infinite "Carousel"

### DISCUSSION THREAD

**Discussion by wesbluemarine** [YaLTeR/niri#45] [Ideas]
I find it "natural" that by pressing **alt+L** on the last window the focus should move "forward" towards the first window. 
What do you think?
Two shortcuts for teleporting to the beginning and end would be also useful.

**Comment by YaLTeR** [comment]
It's a fun idea (and, being a Wayland compositor, nothing prevents niri from drawing an actual infinite carousel this way), but I don't really think it's a natural behavior to have, at least not by default. As for shortcuts for focusing the first and the last window, that's easy to add and I was going to add them at some point anyway.

**Reply by YaLTeR** [reply]
> As for shortcuts for focusing the first and the last window, that's easy to add and I was going to add them at some point anyway.

Added `focus-column-first/last` and `move-column-to-first/last`, bound to Mod+Ctrl?+Home/End by default.

**Comment by rwmpelstilzchen** [comment]
I think there are two ways to display and animate this:

* One is a true infinite carousel, where it seamlessly cycles on itself when you reach either end; for example, say we have `1 2 3` columns, it will show `… 1 2 3 1 2 3 1 2 3 …` and when you move right from `3` to `1` nothing indicates visually you’ve been teleported back, as `1` is also placed ‘right to’ `3`.

* The other is somewhere between the current behaviour and the above. If you have `1 2 3` columns, they show as they currently do (nothing to the right of `3`; nothing to the left of `1`), but calling `focus-column-right` when the focus is on the last column behaves like `focus-column-first` and calling `focus-column-left` on the first column behaves like `focus-column-last`. This option is easier to implement, as it just changes the logic of calling `focus-column-right` from doing nothing on the rightmost column to `focus-column-first` (and analogously the other direction), with no change to the code that manages the visual side of things.

No way is ‘better’ than the other, and IMHO it’s best if both are implemented and the user can choose between a finite carousel (current behaviour), a cyclical carousel (option one) and a ‘[carriage return](https://en.wikipedia.org/wiki/Carriage_return)’ carousel (option two).

**Reply by YaLTeR** [reply]
Someone actually already implemented this: actions `focus-column-right-or-first` and `focus-column-left-or-last`

**Reply by rwmpelstilzchen** [reply]
Great! 🎉
Is this documented anywhere? (I’m not sure, but searching the repo didn’t yield results).

Thank you for Niri! I started using it the day before yesterday and it’s wonderful 🙂

**Reply by YaLTeR** [reply]
Actions aren't documented on the wiki, but they are listed if you run `niri msg action -h`

**Reply by rwmpelstilzchen** [reply]
Mind if I add a page to the wiki, for better discoverability and user-friendliness (I for one, as a user, didn’t find out info about this…)?

**Reply by YaLTeR** [reply]
It's mentioned here: https://github.com/YaLTeR/niri/wiki/Configuration:-Key-Bindings#actions

I just don't want to manually update wiki page with actions every time something changes. Maybe if it would generate automatically (without overwriting the manually written action docs)..

**Reply by rwmpelstilzchen** [reply]
Here’s one-liner that produces the body of a Markdown table which can be pasted into the wiki:

```sh
niri msg action -h | tail -n +6 | head -n -3 | perl -0777 -pe 's/  (\S*)\n          (.*)\n/| `\1` | \2 |\n/g'
```

**Reply by YaLTeR** [reply]
Uhh. I'd much rather extract it programmatically from the parser generated with clap.

**Reply by RGBCube** [reply]
I was wondering about the first way of doing this, how hard would it be to implement? Should I open an issue for it, to see if there is enough demand?

I'd love an actually infinite carousel, similar to how some Android versions let you infinitely cycle between pages in the homescreen.

**Reply by YaLTeR** [reply]
Might turn out complicated, hard to say without trying

**Reply by zaguragit** [reply]
Having the windows infinitely wrap around would actually be really cool for trackpad gesture navigation too, cuz then you have the furthest windows closer to you, and it also fills the empty space left when you choose to always center active windows

**Reply by theNoof** [reply]
whats the intended behaviour when you have 2 windows open that dont fill up the whole screen together? does one of them then get rendered two times?

**Reply by zaguragit** [reply]
Ideally it would be configurable to only wrap around when the total width of all windows exceeds that of the screen..
But, yea, why not?, or maybe to make it less confusing, make them semitransparent, to signify they're not real duplicate windows

**Reply by RGBCube** [reply]
> whats the intended behaviour when you have 2 windows open that dont fill up the whole screen together? does one of them then get rendered two times?

```
+------+ +----+
|      | |    |
|      | |    |
|      | |    |
+------+ +----+
```

\*3 finger swipe left\*


```
-----+ +----+    +--[ <-screen ends here
     | |    |    |  [
     | |    |    |  [
     | |    |    |  [
-----+ +----+    +--[
```

**Reply by theNoof** [reply]
well that would be cool too, thanks for sharing.
